### PR TITLE
[MI-2501] Added unit tests for Carousel input and moved constants to separate package

### PR DESCRIPTION
**Pull request by manojmalik20**
Added unit tests for the HandlePickerSelection API for the carousel section
Added unit tests for new functions in kv.go
Added unit tests for functions in virtual_agent.go file related to carousel input

### CODE REVIEW

File: server/constants/routes.go [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+	PathUserDisconnect             = "/user/disconnect"
+	PathGetUser                    = "/api/now/table/sys_user"
+	PathVirtualAgentWebhook        = "/nowbot/processResponse"
+	PathVirtualAgentBotIntegration = "/api/sn_va_as_service/bot/integration"
+	PathActionOptions              = "/action_options"
+	PathOpenDialog                 = "/api/v4/actions/dialogs/open"
+	PathSetDateTimeDialog          = "/date_time"
+	PathSetDateTime                = "/selected_date_time"
+
+	PathParamEncryptedFileInfo = "encryptedFileInfo"
```
Comment (ayusht2810) [inline server/constants/routes.go]: I think the path param should be in constants.go file

File: server/plugin/kv_test.go [hunk 2]
```diff
@@ -80,3 +81,34 @@ func Test_StoreUser(t *testing.T) {
...
+
+			monkey.Patch(kvstore.LoadJSON, func(_ kvstore.KVStore, _ string, _ interface{}) error {
+				return testCase.loadError
+			})
+
+			_, err := s.LoadPostIDs("mock-userID")
+			if testCase.loadError == nil {
+				require.Nil(t, err)
+			} else {
+				require.NotNil(t, err)
```
Comment (ayusht2810) [inline server/plugin/kv_test.go]: we can check for error value here as well

File: server/plugin/virtual_agent_test.go [hunk 3]
```diff
@@ -452,3 +455,120 @@ func Test_CreateMessageAttachment(t *testing.T) {
...
+func Test_HandleCarouselInput(t *testing.T) {
+	p := Plugin{}
+	for _, test := range []struct {
+		description   string
+		expectedError error
+		setupPlugin   func(p *Plugin, api *plugintest.API)
+		setupStore    func(s *mock_plugin.MockStore)
+	}{
+		{
+			description: "carousel is handled successfully and there is no error",
```
Comment (ayusht2810) [inline server/plugin/virtual_agent_test.go]:
```suggestion
			description: "carousel is handled successfully with no error",
```

File: server/plugin/virtual_agent_test.go [hunk 4]
```diff
@@ -452,3 +455,120 @@ func Test_CreateMessageAttachment(t *testing.T) {
...
+		t.Run(test.description, func(t *testing.T) {
+			mockAPI := &plugintest.API{}
+			test.setupAPI(mockAPI)
+			p.SetAPI(mockAPI)
+			mockCtrl := gomock.NewController(t)
+			mockedStore := mock_plugin.NewMockStore(mockCtrl)
+			test.setupStore(mockedStore)
+			p.store = mockedStore
+
+			p.handlePreviousCarouselPosts(testutils.GetID())
```
Comment (ayusht2810) [inline server/plugin/virtual_agent_test.go:585]: don't we need to check something here?

Comment (manojmalik20) [reply server/plugin/virtual_agent_test.go:585]: No, we don't. If you see the function definition, there's nothing to check. Besides, all the assertions are already happening when we are calling `Expect` function of the mocks.

File: server/plugin/kv_test.go [hunk 5]
```diff
@@ -108,6 +109,7 @@ func Test_LoadPostIDs(t *testing.T) {
 				require.Nil(t, err)
 			} else {
 				require.NotNil(t, err)
```
Comment (ayusht2810) [inline server/plugin/kv_test.go]: we can remove this line now as we are checking the value now

**Review by ayusht2810 — APPROVED** [review]

File: server/plugin/api_test.go [hunk 6]
```diff
@@ -651,6 +676,65 @@ func TestPlugin_handlePickerSelection(t *testing.T) {
 			ParseAuthTokenErr: nil,
 			LoadUserErr:       nil,
 		},
+		"Picker is carousel type and error occurs while getting post": {
```
Comment (Kshitij-Katiyar) [inline server/plugin/api_test.go:679]:
```suggestion
		"Picker is carousel type and the error occurs while getting post": {
```

File: server/plugin/kv_test.go [hunk 7]
```diff
@@ -80,3 +82,35 @@ func Test_StoreUser(t *testing.T) {
...
+	defer monkey.UnpatchAll()
+	for _, testCase := range []struct {
+		description string
+		loadError   error
+	}{
+		{
+			description: "Post IDs are loaded successfully from the KV store",
+		},
+		{
+			description: "Post IDs are not loaded successfully from the KV store",
```
Comment (Kshitij-Katiyar) [inline server/plugin/kv_test.go]:
```suggestion
			description: "Post IDs are not loaded from the KV store",
```

**Review by Kshitij-Katiyar — APPROVED** [review]

File: server/plugin/utils.go [hunk 8]
```diff
@@ -38,8 +44,17 @@ func (p *Plugin) validateTime(time string) string {
...
+		return constants.TimeValidationError
 	}
 
 	return ""
 }
+
+func (p *Plugin) IsCharCountSafe(attachments []*model.SlackAttachment) bool {
+	bytes, err := json.Marshal(attachments)
+	if err != nil {
+		p.API.LogDebug("Error in marshaling the attachments", "Error", err.Error())
```
Comment (avas27JTG) [inline server/plugin/utils.go]: should not we return here in case of error?

File: server/plugin/utils_test.go [hunk 9]
```diff
@@ -114,3 +117,32 @@ func Test_validateTime(t *testing.T) {
...
+	p := Plugin{}
+	for _, test := range []struct {
+		description  string
+		marshalError error
+	}{
+		{
+			description: "result is returned successfully with no error",
+		},
+		{
+			description:  "result is returned successfully with an error",
```
Comment (avas27JTG) [inline server/plugin/utils_test.go]: why is it designed this way that both `result` and `error` are returned?
or may be we can make it more clear with `result is returned as false with an error`

File: server/plugin/virtual_agent_test.go [hunk 10]
```diff
@@ -452,3 +455,120 @@ func Test_CreateMessageAttachment(t *testing.T) {
...
+		},
+		{
+			description: "no postIDs present in KV store",
+			setupAPI:    func(api *plugintest.API) {},
+			setupStore: func(s *mock_plugin.MockStore) {
+				s.EXPECT().LoadPostIDs(testutils.GetID()).Return([]string{}, nil)
+			},
+		},
+		{
+			description: "error in storing postIDs and getting post from API",
```
Comment (avas27JTG) [inline server/plugin/virtual_agent_test.go]:
```suggestion
			description: "error occurred while storing postIDs and getting post from API",
```
I think `while` sounds more clear than `in`, let me know your opinion

File: server/plugin/utils.go (hunk 8 above)
Comment (manojmalik20) [reply server/plugin/utils.go]: I don't even think that there's a need to handle the error here, so reverting this.

**Review by avas27JTG — APPROVED** [review]

**Pull request merged**